src/librustc_data_structures/stable_map.rs

```diff
@@ -0,0 +1,80 @@
+pub use rustc_hash::FxHashMap;
+use std::borrow::Borrow;
+use std::collections::hash_map::{Entry, IterMut};
+use std::hash::Hash;
+
+/// A deterministic wrapper around FxHashMap that does not provide iteration support.
+///
+/// It supports insert, remove, get and get_mut functions from FxHashMap.
+/// It also allows to convert hashmap to a sorted vector with the function `into_sorted_vector()`.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct StableMap<K, V>
+where
+    K: Eq + Hash,
```

Comment:
This is not strictly necesary.

Reply:
If I do not do this, I get the errors for `Eq` and `Hash` trait bounds not being satisfied. Am I doing something incorrectly?

Reply:
Did you keep them on the impl block?

Reply:
Yes.

Reply:
Strange. Doesn't hurt to keep this where though.

Reply:
Actually, it does hurt -- the bounds here mean that any generic struct with a StableMap inside it needs to also list the bounds. They shouldn't be necessary unless something is wrong; could you list the error message?

Reply:
Here is the log for `Debug` with `Eq` trait. There is a long log like this for every derived attribute and `Eq` and `Hash` traits.
```
error[E0277]: the trait bound `K: std::cmp::Eq` is not satisfied
  --> src/librustc_data_structures/stable_map.rs:13:5
   |
10 | #[derive(Debug, Clone, Default, PartialEq, Eq)]
   |          ----- in this macro invocation
...
13 |     base: FxHashMap<K, V>,
   |     ^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `K`
   | 
  ::: <::core::fmt::macros::Debug macros>:1:1
   |
1  | ($ item : item) => { }
   | ---------------------- in this expansion of `#[derive(Debug)]`
   |
   = help: consider adding a `where K: std::cmp::Eq` bound
   = note: required because of the requirements on the impl of `std::fmt::Debug` for `std::collections::HashMap<K, V, std::hash::BuildHasherDefault<rustc_hash::FxHasher>>`
   = note: required because of the requirements on the impl of `std::fmt::Debug` for `&std::collections::HashMap<K, V, std::hash::BuildHasherDefault<rustc_hash::FxHasher>>`
   = note: required for the cast to the object type `dyn std::fmt::Debug`
```
Let me know if full log would be rather useful, shall attach a file.

Reply:
Ah, I see the problem, you can't derive Debug because it's insufficiently smart. If you write a simple Debug impl yourself we should be able to avoid the bounds issue on the struct itself

Reply:
Using an own `Debug` implementation would also make the debug output prettier, as deriving it would give `StableMap { base: { 1: 2, 3: 4 } }` instead of `{ 1: 2, 3: 4 }`.

Reply:
Hi @bjorn3 and @Mark-Simulacrum !
I wrote a Debug impl but now the error exists for `Eq`,  `PartialEq` since `Eq` is bound by `PartialEq` and `Hash`. I believe I will have to write their impl too? But that also makes me wonder why didn't the other map data structures need this?

Reply:
I'm not sure what other map data structures you're talking about, but yeah, reimplementing all of those seems like it'd probably be necessary.

We probably dont' want to implement `Hash` as it's not generally implemented for Hash{Map,Set}.